general/error.hpp

```diff
@@ -151,6 +151,11 @@ void mfem_warning(const char *msg = NULL);
       printf(msg);             \
       asm("trap;");            \
```

Comment:
The above `asm("trap;");` line didn't work at all with, HIP, correct?

Reply:
cc: @dylan-copeland, @artv3

Reply:
Yes, that's correct.  I tried some variations on the compiler's suggested replacement (`s_trap`, I believe it was) but I couldn't get one that worked, so I just removed the line.

Reply:
For hip abort should work with device code:
``` 
#if defined(__HIP_DEVICE_COMPILE__)
  abort();
#endif
```

Reply:
Thanks @artv3. Is `__HIP_DEVICE_COMPILE__` something that should be detected by my HIP configuration, or something I should set when building?  When I tried adding the above code, I got the same error as before I added anything related to HIP (error because it's trying to call the standard MFEM abort, so it says it's trying to call a host function from the device) which leads me to believe it didn't think `__HIP_DEVICE_COMPILE__` was defined

Reply:
``__HIP_DEVICE_COMPILE__`` should be automatically defined when the compiler is compiling for device code. 

See: Identifying Current Compilation Pass: Host or Device in https://rocmdocs.amd.com/en/latest/Programming_Guides/HIP-porting-guide.html

Reply:
Hmm, I'm not sure what's going on.  I keep getting errors like 
`mfem/linalg/sparsemat.cpp:2370:16: error: reference to __host__ function 'mfem_warning' in __device__ function`

Looks like it's coming from an `MFEM_ABORT` call inside `MFEM_FORALL`

Reply:
Was this on Corona?

Reply:
No, on a local machine here at ICL.

Reply:
What's going on here is that when hip-clang does its host and compile passes all of the code is ALWAYS visible. So when it is doing the host compilation pass __HIP_DEVICE_COMPILE__ is not defined, but it still checks that the __device__ functions are sane in that situation. This makes using macros much more difficult. 

In this case when doing the host pass but looking at the device code, the check in the macro for #defined(__HIP_DEVICE_COMPILE__) says no, and it goes on to try and place the following chain of macros into __device__ code:

#define MFEM_ABORT_KERNEL(msg) MFEM_ABORT(msg) 
#define MFEM_ABORT(msg) _MFEM_MESSAGE("MFEM abort: " << msg, 0)
#define _MFEM_MESSAGE(msg, warn)                                        \
   {                                                                    \
      std::ostringstream mfemMsgStream;                                 \
      mfemMsgStream << std::setprecision(16);                           \
      mfemMsgStream << std::setiosflags(std::ios_base::scientific);     \
      mfemMsgStream << msg << MFEM_LOCATION;                            \
      if (!(warn))                                                      \
         mfem::mfem_error(mfemMsgStream.str().c_str());                 \
      else                                                              \
         mfem::mfem_warning(mfemMsgStream.str().c_str());               \
   }


and obviously the MFEM_MESSAGE macro contains lots of host code the GPU doesn't understand. 

A way to get this to work with hip clang would be something like this:


#if defined(MFEM_USE_HIP)
template <typename T>
__host__ void abort_msg(T & msg)
{
   MFEM_ABORT(msg);
}

template <typename T>
__device__ void abort_msg(T & msg)
{
   abort();
}

#else
template <typename T>
__host__ __device__ void abort_msg(T & msg)
{
#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE)
#else
   MFEM_ABORT(msg);
#endif
}
#endif
// Abort inside a device kernel
#if defined(__CUDA_ARCH__)
#define MFEM_ABORT_KERNEL(msg) \
   {                           \
      printf(msg);             \
      asm("trap;");            \
   }
#else
#define MFEM_ABORT_KERNEL(msg) \
   {                           \
      abort_msg(msg);          \
   }
#endif
#endif


You have to separate the nvcc and hip-clang cases because with nvcc you cannot have __host__ and __device__ functions with the same function signature (with clang the __host__ and __device__ tags are part of the function signature but they aren't with nvcc).

Reply:
@moes1, thanks for the explanation.   In f2d5e1a1d3768864aef09c0778f324207b1b8a26 I tried incorporating this in a way that had fewer changes to what was in MFEM currently -- but I'm not entirely sure it is functionally the same as what you suggested.  And, if I understand correctly, if MFEM-HIP aborts from inside a device kernel, it still won't print the error message, unlike MFEM-CUDA?  Or no?

Reply:
Yes I removed the printf from the __device__ abort_msg implementation. You could add it and then it would print from within the kernel when it aborts. However I think the recommended practice then would be to call it from the abort_msg function decorated with "__attribute__((noinline))" because if the compiler attempts to inline the printf calls it can cause problems. 

So:

__attribute__((noinline)) __device__ void abort_msg(T & msg)
{
   printf(msg);
   abort();
}

Reply:
Actually after further thought, I would recommend not using printf with hip unless debugging if possible right now...